File: src/main/java/com/github/svegon/capi/event/render/StatusEffectLightCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.network.ClientPlayerEntity;
import net.minecraft.entity.effect.StatusEffect;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface StatusEffectLightCallback {
    Event<StatusEffectLightCallback> EVENT = EventFactory.createArrayBacked(StatusEffectLightCallback.class,
            (player, effect, callback) -> {}, listeners -> (player, effect, callback) -> {
        for (StatusEffectLightCallback listener : listeners) {
            listener.onLightStatusEffectCheck(player, effect, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void onLightStatusEffectCheck(ClientPlayerEntity player, StatusEffect effect,
                                  CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/StatusEffectLightCallback.kt
package io.github.svegon.capi.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.network.ClientPlayerEntity
import java.util.function.Function

interface StatusEffectLightCallback {
    fun onLightStatusEffectCheck(
        player: ClientPlayerEntity?, effect: StatusEffect?,
        callback: CallbackInfoReturnable<Boolean?>?
    )

    companion object {
        @JvmField
        val EVENT: Event<StatusEffectLightCallback?> = EventFactory.createArrayBacked<StatusEffectLightCallback?>(
            StatusEffectLightCallback::class.java,
            StatusEffectLightCallback { player: ClientPlayerEntity?, effect: StatusEffect?, callback: CallbackInfoReturnable<Boolean?>? -> },
            Function<Array<StatusEffectLightCallback?>, StatusEffectLightCallback?> { listeners: Array<StatusEffectLightCallback?>? ->
                StatusEffectLightCallback { player: ClientPlayerEntity?, effect: StatusEffect?, callback: CallbackInfoReturnable<Boolean?>? ->
                    for (listener in listeners) {
                        listener.onLightStatusEffectCheck(player, effect, callback)

                        if (callback.isCancelled()) {
                            return@StatusEffectLightCallback
                        }
                    }
                }
            })
    }
}
